### PR TITLE
Fix units from transports appearing at load point.

### LINKED ISSUE #16901: Units moving into the world from a transport appear at the load point
They then blink to the unload point once the move animation has completed.

See https://forum.openra.net/viewtopic.php?p=310600#p310600

### PR DESCRIPTION
**Pull request by tovl**
Fixes #16901

The `MoveIntoWorld` activity must not be interruptible and `Enter` should always be allowed to exit gracefully on its own because that would otherwise cause other glitches. So instead make sure that `Enter` chooses the current location for exiting and just let that handle the unload movement.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/EnterTransport.cs [hunk 1]
```diff
@@ -63,10 +63,6 @@ protected override void OnEnterComplete(Actor self, Actor targetActor)
 
 				enterCargo.Load(enterActor, self);
 				w.Remove(self);
-
-				// Preemptively cancel any activities to avoid an edge-case where successively queued
-				// EnterTransports corrupt the actor state. Activities are cancelled again on unload
-				self.CancelActivity();
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/EnterTransport.cs]: We will still need this to cancel any activities queued after the `EnterTransport` (e.g. backup orders to enter a different transport if the first one was full).

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 2]
```diff
@@ -121,9 +121,12 @@ public override bool Tick(Actor self)
 					var move = actor.Trait<IMove>();
 					var pos = actor.Trait<IPositionable>();
 
-					actor.CancelActivity();
+					pos.SetPosition(self, exitSubCell.Value.First, exitSubCell.Value.Second);
 					pos.SetVisualPosition(actor, spawn);
-					actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
+					if (actor.CurrentActivity == null)
+						actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: The problem at hand is really an implementation issue/detail of `EnterTransport`, so it would be a lot cleaner if we could fix the issue in `EnterTransport`/`Enter` instead of adding workarounds (especially uncommented ones!) in other parts of the code.

Experience shows us that external workarounds tend to be forgotten when the code they are working around is inevitably rewritten, causing confusion and headaches for people in the future who are trying to look back and work out what odd code snippets were originally trying to solve.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/UnloadCargo.cs]: One simple(?) option could be to change `OnEnterComplete` to return a bool: returning false keeps the current behaviour where it transitions to `EnterState.Exiting`,  returning true jumps directly to `EnterState.Complete`.

File: OpenRA.Mods.Common/Activities/EnterTransport.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Mods.Common/Activities/EnterTransport.cs]: ...err maybe not? It would help if I read the comment above before making my own comments. Still, we should make sure to test that this still behaves correctly.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/EnterTransport.cs]: This `CancelActivity` is not needed because we already cancel the activity queue in `UnloadCargo`. Cancelling the activity queue twice smells of a hack. It does nothing and if it did it would be a bug.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs (hunk 2 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/UnloadCargo.cs]: This is not adding an external workaround but removing one.

`Enter` and all its derived classes always take care of exiting as well. This makes sense because everything that enters must at some point exit (or despawn) before it does anything else. Whether it has finished or cancelled, the `MoveIntoWorld` must always happen and cannot be interrupted. `EnterTransport` isn't (and should be) and exception. Anything that enters a transport must exit before it can do anything else. The only difference is that the place where it exits is not the same as where it entered.

Implementing complex exceptions in `Enter` is a bad idea: It makes the behaviour contradictory with every other use case of `Enter` and makes the activity even more convoluted than it already is. I predict that this approach is much more susceptible to unforeseen regressions.

We would also be shooting ourselves in the foot in the long run: The way the activity system is designed, we cannot inject an activity in the middle of a queue (nor should we feel the need to) and we must avoid messing with running activity queues whenever possible. This means it would be impossible to implement for example an autotransport (analogous with `AutoCarryall`) if we would go this route. On the basis of this PR we would only need to make the last `actor.CancelActivity` call optional to make that possible.

Furthermore, any cargo actor that we can unload must have either entered at some point or spawned there. So we only need to manually queue `MoveIntoWorld` in the latter case (which only happens with scripted transports or a hypothetical future `ProductionTransport` trait). `UnloadCargo` should ideally not touch the cargo's activity queue at all but simply drop the cargo into the world with a specified position and let the cargo take care of the exit animation itself in all cases. We could do that by letting `Mobile` queue a `MoveIntoWorld` activity upon creation (which would also clean op the `Production` traits somewhat). Modelling `UnloadCargo` this way would also allow us to merge `UnloadCargo` and `DeliverUnit` quite easily, because `DeliverUnit` already operates on that model.

So, TL;DR: I think my solution is more internally consistent and sustainable in the long run.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/UnloadCargo.cs]: This will work better if we can rename `EnterTransport` to make it clear that it also manages exit, and ideally also use the same activity for units spawned in the transport instead of having a separate code path for `MoveIntoWorld` here.

**Comment by tovl** [comment]
Renamed `EnterTransport` to `RideTransport` and removed all calls to `MoveIntoWorld` from `UnloadCargo` as well as all `Production` traits.

`MoveIntoWorld` is now always called on actor creation instead and is changed to be a proper (noninterruptible) activity. This seems to have solved #16930 as well.

**Comment by pchote** [comment]
The transport changes look reasonable, but forcing the `MoveIntoWorld` activity for all actors regresses probably most non-production things that create actors.

Aug-16-2019 17-59-39

Notice how this makes the paradrop plane do a loop when entering the map, and the dropped actors not open their parachutes until after they touch down.

This could be fixed by changing `DelayInit` into a `MoveIntoWorld` init, and only queue the activity if it is explicitly set.

**Comment by tovl** [comment]
Fixed the issues with aircraft and paratroopers. Note that these issues where not actually exclusive to non-production cases. The `MoveIntoWorld` method for aircraft just generally didn't make much sense before (and wasn't used either), so I changed it to cover the actions that an aircraft actually would do when moving into the world.

File: OpenRA.Mods.Common/ActorInitializer.cs [hunk 3]
```diff
@@ -24,6 +24,16 @@ public class FacingInit : IActorInit<int>
 		public int Value(World world) { return value; }
 	}
 
+	public class DelayInit : IActorInit<int>
```
Comment (pchote) [inline OpenRA.Mods.Common/ActorInitializer.cs]: `DelayInit` is too generic and may conflict with other Inits in the future (also don't forget that these are automatically bridged to `map.yaml`). Please rename this to `MoveIntoWorldDelayInit` to make its use clear, even if you don't want to explicitly gate the activity on it existing.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 4]
```diff
@@ -219,6 +219,7 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 		public bool ForceLanding { get; private set; }
 		IEnumerable<CPos> landingCells = Enumerable.Empty<CPos>();
 		bool requireForceMove;
+		int initDelay;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Likewise here - please call this `moveIntoWorldDelay` to make its meaning immediately clear.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 5]
```diff
@@ -868,9 +857,51 @@ public Activity MoveTo(CPos cell, Actor ignoreActor, Color? targetLineColor = nu
...
+
+				if (!aircraft.Info.TakeOffOnCreation)
+					return;
+
+				// Move to the host's rally point if it has one
+				var rp = aircraft.ReservedActor != null ? aircraft.ReservedActor.TraitOrDefault<RallyPoint>() : null;
+				if (self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition).Length <= aircraft.LandAltitude.Length)
+				{
+					if (rp != null)
+						QueueChild(new TakeOff(self, Target.FromCell(self.World, rp.Location)));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: If the `Wait` above is too long `rp.Location` may change between the time this is queued and the activity actually ticks. It would be better to defer this until the first `Tick`, but we'll need to be careful that the host hasn't died in the meantime.

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 6]
```diff
@@ -144,6 +144,7 @@ public class Mobile : PausableConditionalTrait<MobileInfo>, IIssueOrder, IResolv
 	{
 		readonly Actor self;
 		readonly Lazy<IEnumerable<int>> speedModifiers;
+		readonly int initDelay;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: Likewise here.

File: OpenRA.Test/OpenRA.Test.csproj [hunk 7]
```diff
@@ -45,4 +45,4 @@
       <Analyzer Remove="@(Analyzer)" />
     </ItemGroup>
   </Target>
-</Project>
\ No newline at end of file
+</Project>
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Test/OpenRA.Test.csproj]: This is still an issue.

**Review by pchote — COMMENTED** [review]
My gut feeling is still that always forcing the `MoveIntoWorld` activity is going to cause problems, but (a) I can't think of any specific examples off the top of my head and (b) fixing this by only queueing the activity if the Init is defined is a simple future fix. I don't see much point wasting both our time banging heads over that without a firmer basis - but please address the naming nits below:

**Comment by tovl** [comment]
Applied fix-ups and updated authors list.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 8]
```diff
@@ -219,6 +219,7 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 		public bool ForceLanding { get; private set; }
 		IEnumerable<CPos> landingCells = Enumerable.Empty<CPos>();
 		bool requireForceMove;
+		int moveIntoworldDelay;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Typo: `moveIntoWorldDelay` (capital W)

**Review by pchote — COMMENTED** [review]
Code changes look reasonable aside from one small typo. I'll try to get to test this later tonight.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 8 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed typo.

File: OpenRA.Mods.Common/Traits/Crates/Crate.cs [hunk 9]
```diff
@@ -94,6 +95,11 @@ public Crate(ActorInitializer init, CrateInfo info)
 				SetPosition(self, init.Get<LocationInit, CPos>());
 		}
 
+		void INotifyCreated.Created(Actor self)
+		{
+			self.QueueActivity(new Parachute(self));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crates/Crate.cs]: This crashes in mods that don't have parachutable crates (anything but RA).

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Crates/Crate.cs]: This should really be handled the same way as paratroopers.

File: OpenRA.Mods.Common/Traits/Crates/Crate.cs [hunk 10]
```diff
@@ -94,6 +95,12 @@ public Crate(ActorInitializer init, CrateInfo info)
 				SetPosition(self, init.Get<LocationInit, CPos>());
 		}
 
+		void INotifyCreated.Created(Actor self)
+		{
+			if (self.TraitOrDefault<Parachutable>() != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crates/Crate.cs]: Please also include the height check here, otherwise crates spawned on the ground (e.g. destroyed supply trucks) will blink the parachute opening animation.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Crates/Crate.cs]: Fixed.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 11]
```diff
@@ -868,9 +857,53 @@ public Activity MoveTo(CPos cell, Actor ignoreActor, Color? targetLineColor = nu
...
-			return new Fly(self, Target.FromCell(self.World, cell, subCell));
+			readonly Actor self;
+			readonly Aircraft aircraft;
+			readonly int delay;
+
+			public MoveIntoWorldActivity(Actor self, int delay = 0)
+			{
+				this.self = self;
+				aircraft = self.Trait<Aircraft>();
+				IsInterruptible = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: If the actor can't make it to the rallypoint (e.g. is blocked by repulsion) this causes the helicopter to be stuck uncontrollable. Simple repro case: enable cheats and build five helicopters at once.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed as discussed on IRC

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
This seems to have broken the shellmap, and I guess other scripts that script units loaded into transports (reported from the latest test build).

**Comment by pchote** [comment]
False alarm: with my sped up test case i've been able to reproduce this on bleed, so can't be a regression here.

**Review by abcdefg30 — COMMENTED** [review]
Looks good to me otherwise.

**Comment by tovl** [comment]
Style nits fixed.

> False alarm: with my sped up test case i've been able to reproduce this on bleed, so can't be a regression here.

:disappointed_relieved:

**Comment by pchote** [comment]
Note: double false alarm, because the same issue also exists in release-20190314 and earlier 😌

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4ad27729d5d2a151e62a98790861db862d5488a6...673f3ddf3588a3ca429c9e98368bb4e0a825bdba)
